src/is_wire/rpc/service_provider.py

```diff
@@ -0,0 +1,98 @@
+from ..core import Channel, Subscription, Status, StatusCode, Logger
+from ..core.utils import assert_type
+from .context import Context
+import traceback
+import six
+from google.protobuf.json_format import ParseError
+
+
+class ServiceProvider(object):
+    log = Logger("ServiceProvider", Logger.DEBUG)
+
+    def __init__(self, channel):
+        assert_type(channel, Channel, "channel")
+        self._channel = channel
+        self._services = {}
+        self._interceptors_before = []
+        self._interceptors_after = []
+
+    def delegate(self, topic, function, request_type, reply_type):
+        """ Bind a function to a particular topic, so everytime a message is
+            received in this topic the function will be called """
+        assert_type(topic, six.string_types, "topic")
+        self.log.debug("New service registered '{}'", topic)
+        subscription = Subscription(self._channel, name=topic)
+        wrapped = self.wrap(function, request_type, reply_type)
+        self._services[subscription.id] = wrapped
```

Comment:
If the user delegate two different functions to the same service, i.e.:
```python
provider.delegate('MyService', impl1, RequestType, ReplyType)
provider.delegate('MyService', impl2, RequestType, ReplyType)
```
the behaviour is like have two instances of the same service. Could be interesting, in case of already have a handler to the delegated service, re-bind it (overwriting the previous implementation) and generate a message in order to advertise the user.

Reply:
Chaging the implementation of a service does not seem like common/desireable behaviour. So I would rather raise an exception when redelegating on the same topic.